PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsInfoProvider.kt

```diff
@@ -59,7 +61,28 @@ object FunctionsCoreToolsInfoProvider {
             }
         }
 
-        return FunctionsCoreToolsInfo(funcCoreToolsPath, coreToolsExecutablePath.path)
+        return FunctionsCoreToolsInfo(funcCoreToolsPath.path, coreToolsExecutablePath.path)
+    }
+
+    private fun patchCoreToolsPath(funcCoreToolsPath: File): File {
+
+        if (!SystemInfo.isWindows) return funcCoreToolsPath
+
+        // Check if it's a Chocolatey install or NPM install.
+        // If it is, rewrite the path to the tools path where the func executable is located.
+        val chocolateyCoreToolsPath = funcCoreToolsPath.resolve("..").resolve("lib").resolve("azure-functions-core-tools").resolve("tools")
+        if (chocolateyCoreToolsPath.exists()) {
+            logger.info("Functions core tools path ${funcCoreToolsPath.path} is Chocolatey-installed. Rewriting path to ${chocolateyCoreToolsPath.path}")
+            return chocolateyCoreToolsPath
+        }
+
+        val npmCoreToolsPath = funcCoreToolsPath.resolve("..").resolve("node_modules").resolve("azure-functions-core-tools").resolve("bin")
+        if (npmCoreToolsPath.exists()) {
+            logger.info("Functions core tools path ${funcCoreToolsPath.path} is NPM-installed. Rewriting path to ${npmCoreToolsPath.path}")
+            return npmCoreToolsPath
+        }
+
+        return funcCoreToolsPath
```

Comment:
Maybe check for current path existence first? So, we do not need to make this extra checks if main path exists.

Also, I'm not quite sure I understand the logic here - we take a value that user pass in settings and instead of validating this value in settings we modify it silently. Maybe, it would be better to make this path setup logic on the Settings level?

Reply:
This is essentially what is was before. A path check is done [here](https://github.com/JetBrains/azure-tools-for-intellij/pull/527/files/e372924c30015eb4cf5d36727c46f169a2918c5f#diff-c6d6d9a6e6734862278125aebfe3c5e5d197dd9508b025e2053bbff9a94f5f22R39) already.

As for the logic, looks like I missed that in the PR opening comment, will add it there for future reference as well. Here goes:

Chocolatey and NPM installs are just shims. The executable starts the actual `func.exe` as a child process. Our debugger tries to attach to the shim process, which is a native executable, and that will always fail. Finding the actual `func.exe` and running that instead solves this issue.

* We auto-detect the functions core tools path, using [the logic](https://github.com/JetBrains/azure-tools-for-intellij/pull/527/files/e372924c30015eb4cf5d36727c46f169a2918c5f#diff-c6d6d9a6e6734862278125aebfe3c5e5d197dd9508b025e2053bbff9a94f5f22R104) from the upstream Microsoft version of the plugin
  * This logic checks `PATH`, and resolves the full path to the core tools
  * If the full path is Chocolatey or NPM, similar logic to ours here is applied to patch it to the correct path
* This PR works just before run: we patch the path if it's required for the debugger to work. Agree on moving it into the settings dialog instead and will fix that.